### PR TITLE
Ananth Bhaskararaman's solution

### PR DESCRIPTION
**Pull request by ananthb**
My solution is written in golang.

It can be built with `go build` and executed with
`./schedule -problem 1` to solve the first problem and
`./schedule -problem 2` to solve the second problem.

### CODE REVIEW

**Comment by maknahar** [comment]
Hi @ananthb, Thanks for submitting the solution. The solution to the first problem seems perfect. However, Solution of the second problem seems to output different output than expected. I am attaching the sample input file content:

capacities:

```
"Partner ID","Capacity (in GB)"
P1          ,800
P2          ,800
P3          ,800
```

input:

```
D2,320,T1
D1,350,T1
D3,510,T1
D4,700,T2
D5,300,T5
```

Partners

```
Theatre,Size Slab (in GB),Minimum cost,Cost Per GB,Partner ID
T1     ,0-100            ,1500        ,20         ,P1
T1     ,100-200          ,1900        ,13         ,P1
T1     ,200-300          ,2500        ,12         ,P1
T1     ,300-400          ,3000        ,10         ,P1
T1     ,200-400          ,2500        ,15         ,P2
T1     ,100-200          ,800         ,25         ,P3
T1     ,200-600          ,1200        ,30         ,P3
T2     ,0-100            ,1500        ,20         ,P1
T2     ,100-200          ,2000        ,15         ,P1
T2     ,200-300          ,2500        ,12         ,P1
T2     ,300-400          ,3000        ,15         ,P1
T2     ,0-200            ,2500        ,20         ,P2
T2     ,200-400          ,3500        ,10         ,P2
T2     ,100-200          ,900         ,15         ,P3
T2     ,200-400          ,1000        ,12         ,P3
```

**Comment by maknahar** [comment]
Output that I am getting is:

```
D1,false,,
D2,false,,
D3,true,P3,15300
D4,false,,
D5,false,,
```

Expected:
```
D4,false,,
D3,true,P3,15300
D1,true,P1,3500
D2,true,P2,4800
D5,false,,
```

Order may changes.

**Comment by ananthb** [comment]
Hey @maknahar, I believe I've fixed it. My solution is slightly different as it works out cheaper than the one in your expected output. Could you test this please?

File: schedule.go [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+	)
+	flag.IntVar(&problemNumber, "problem", defaultProblemNumber, "Choose between the two problems")
+	flag.StringVar(&partnersFile, "partners", defaultPartnersFile, "Specify a partners list")
+	flag.StringVar(&capacitiesFile, "capacities", defaultCapacitiesFile, "Specify capacities")
+	flag.StringVar(&inputFile, "input", defaultInputFile, "Input")
+}
+
+func readPartners() []Partner {
+	partnersCSV, err := os.Open(partnersFile)
+	defer partnersCSV.Close()
```
[the remark this answers is not in the document]
Comment (ananthb) [reply schedule.go]: Defer statements execute when the function they are in return. Besides aesthetics, my understanding is that the program will behave the same when the defer is before the error checking or after.

File: schedule.go [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+	capCSV, err := os.Open(capacitiesFile)
+	defer capCSV.Close()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	cr := csv.NewReader(capCSV)
+	capacities := make(Capacity)
+	cr.Read() // dump header
+	for {
```
[the remark this answers is not in the document]
Comment (ananthb) [reply schedule.go]: I could take the record reading portion of it and re-write it like:

```
func readRecord(csv.Reader) []string {
    ...
}
```

**Comment by ananthb** [comment]
@maknahar is this what you were looking for?

**Comment by maknahar** [comment]
@ananthb Check these input once?

capacities:

```
"Partner ID","Capacity (in GB)"
P1          ,350
P2          ,300
P3          ,800
```

input:

```
D1,370,T1
D2,325,T2
D3,510,T1
D4,700,T2
D5,300,T5
```

Partners

```
Theatre,Size Slab (in GB),Minimum cost,Cost Per GB,Partner ID
T1     ,0-100            ,1500        ,20         ,P1
T1     ,100-200          ,1900        ,13         ,P1
T1     ,200-300          ,2500        ,12         ,P1
T1     ,300-400          ,3000        ,10         ,P1
T1     ,200-400          ,2500        ,15         ,P2
T1     ,100-200          ,800         ,25         ,P3
T1     ,200-600          ,1200        ,30         ,P3
T2     ,0-100            ,1500        ,20         ,P1
T2     ,100-200          ,2000        ,15         ,P1
T2     ,200-300          ,2500        ,12         ,P1
T2     ,300-400          ,3000        ,15         ,P1
T2     ,0-200            ,2500        ,20         ,P2
T2     ,200-400          ,3500        ,10         ,P2
T2     ,100-200          ,900         ,15         ,P3
T2     ,200-400          ,1000        ,12         ,P3
```

**Comment by ananthb** [comment]
If i run it as the first problem, I get:

```csv
D1,true,P1,3700
D2,true,P2,3500
D3,true,P3,15300
D4,false,,
D5,false,,
```
and as problem two I get,

```csv
D1,true,P1,3700
D2,true,P1,4875
D3,true,P3,15300
D4,false,,
D5,false,,
```

**Comment by maknahar** [comment]
Solution for problem 2 is where is the issue is. Does P1 has optimum capacity?

**Comment by ananthb** [comment]
I just noticed that as well. P1 is over-booked in this instance. Let me check that out real quick.

**Comment by ananthb** [comment]
It turned out to be a pretty simple fix. Could you test it now please?

**Pull request closed without merging**